### PR TITLE
[Enchance] internlm message to prompt

### PR DESCRIPTION
**Pull request by Harold-lkk**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: lmdeploy/model.py [hunk 1]
```diff
@@ -368,15 +372,16 @@ def messages2prompt(self, messages, sequence_start=True):
...
-                ret += f'{self.user}{user}{self.eoh}{self.assistant}' \
-                       f'{assistant}{self.eoa}'
-            else:
-                ret += f'{self.user}{user}{self.eoh}{self.assistant}'
+        eox_map = dict(user=self.eoh, assistant=self.eoa, system=self.eosys)
+        ret = '<BOS>'
+        if self.meta_instruction:
+            ret += f'{self.system}{self.meta_instruction}{self.eosys}'
+
+        for message in messages:
```
Comment (lvhan028) [inline lmdeploy/model.py:380]: What if one of the messages is `system:content`?

File: lmdeploy/model.py [hunk 2]
```diff
@@ -207,12 +209,12 @@ def decorate_prompt(self, prompt, sequence_start=True):
         assert self.capability == 'chat', \
             f'{type(self).__name__} has no capability of {self.capability}'
         if sequence_start:
-            return f'<BOS>{self.system}:{self.meta_instruction}\n' \
-                   f'{self.user}:{prompt}{self.eoh}\n' \
-                   f'{self.assistant}:'
+            return f'<BOS>{self.system}{self.meta_instruction}{self.eosys}' \
```
Comment (AllentDan) [inline lmdeploy/model.py:212]: why not rename `system` to `bosys`?

File: lmdeploy/model.py [hunk 3]
```diff
@@ -207,12 +209,12 @@ def decorate_prompt(self, prompt, sequence_start=True):
...
+                   f'{self.user}{prompt}{self.eoh}' \
+                   f'{self.assistant}'
         else:
-            return f'\n{self.user}:{prompt}{self.eoh}\n' \
-                   f'{self.assistant}:'
+            return f'\n{self.user}{prompt}{self.eoh}' \
```
Comment (AllentDan) [inline lmdeploy/model.py:216]: do we need `\n` here? As `eoa='<eoa>\n'` now.

File: lmdeploy/model.py (hunk 2 above)
Comment (Harold-lkk) [reply lmdeploy/model.py:212]: keep same with self.user/assistant

File: lmdeploy/model.py (hunk 3 above)
Comment (Harold-lkk) [reply lmdeploy/model.py:216]: but stop  word not  contain `\n`

Comment (AllentDan) [reply lmdeploy/model.py:216]: So, `eoa` variable can't be a stop word here. This may make user confused. 🤔

**Comment by AllentDan** [comment]
I found we both implemented the chat template differently from the huggingface codes.
https://huggingface.co/internlm/internlm-chat-7b-v1_1/blob/main/modeling_internlm.py#L769

**Comment by Harold-lkk** [comment]
> I found we both implemented the chat template differently from the huggingface codes. https://huggingface.co/internlm/internlm-chat-7b-v1_1/blob/main/modeling_internlm.py#L769

It's wrong, and will be fixed

**Comment by AllentDan** [comment]
@Harold-lkk Please provide some example messages containing two or more system roles. Tested OK with single system role.

**Review by AllentDan — APPROVED** [review]
LGTM

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**